File: perf/benchmarks/micro/sort/sort.py
# Sort N integers — stdlib-sort micro-benchmark.
#
# We generate a deterministic LCG sequence, then time only the sort step.
# Argument: N (default 2_000_000).

import sys
import time


def gen_data(n: int) -> list[int]:
    seed = 12345
    arr = [0] * n
    for i in range(n):
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        arr[i] = seed
    return arr


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 2_000_000
    arr = gen_data(n)
    t0 = time.perf_counter_ns()
    arr.sort()
    elapsed_ns = time.perf_counter_ns() - t0
    checksum = arr[0] ^ arr[n // 2] ^ arr[-1]
    print(f"CHECKSUM:{checksum}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()

File: perf/benchmarks/micro/sort/sort.cpp
// Sort N integers — stdlib-sort micro-benchmark (C++).
#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 2'000'000LL;
    std::vector<int64_t> arr(static_cast<size_t>(n));
    int64_t seed = 12345;
    for (int64_t i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        arr[static_cast<size_t>(i)] = seed;
    }
    auto t0 = std::chrono::steady_clock::now();
    std::sort(arr.begin(), arr.end());
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    int64_t cs = arr.front() ^ arr[static_cast<size_t>(n / 2)] ^ arr.back();
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}
